lib/puppet/property.rb

```diff
@@ -372,17 +372,27 @@ def insync?(is)
   # @return [Boolean] whether or not the values are in sync or not
   # @api private
   def insync_values?(should, is)
-    # Here be dragons. We're duplicating the resource property itself so we can pass
-    # a new should value without side-effect. This kind of mechanism is undesirable,
-    # but it avoids the risk of causing side-effects on live properties somewhere
-    # else in code. The duplicate should be short-lived and is only required for this
-    # comparison so it can be GC'd directly afterwards. Unfortunately there isn't
-    # an api compatible way of avoiding this, as both should and insync? behaviours
-    # are part of the API. Future API work should factor this kind of arbitrary comparison
-    # into the API to remove this complexity. -ken
-    test_property = self.dup
-    test_property.should = should
-    test_property.insync?(is)
+    # Here be dragons. We're setting the should value of a property purely just to
+    # call its insync? method, as it lacks a way to pass in a should.
+    # Unfortunately there isn't an API compatible way of avoiding this, as both should
+    # an insync? behaviours are part of the public API. Future API work should factor
+    # this kind of arbitrary comparisons into the API to remove this complexity. -ken
+    begin
+      # Backup old should, set it to the new value, then call insync? on the property.
+      old_should = self.should
+      self.should = should
+      result = insync?(is)
+
+      # Restore old should, and return results from insync?
+      self.should = old_should
```

Comment:
This needs to be in an ensure block or we could leave the property with a bad should value